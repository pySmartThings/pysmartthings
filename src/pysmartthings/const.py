"""Define consts for the pysmartthings package."""

API_BASE = "api.smartthings.com"
# Version required to use SSE
API_VERSION = 20250122
