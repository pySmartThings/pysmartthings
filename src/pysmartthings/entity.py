"""Defines the entity module."""

from typing import NoReturn

from .api import Api


class Entity:
    """Define an entity from the SmartThings API."""

    def __init__(self, api: Api) -> None:
        """Initialize a new instance of the entity."""
        self._api = api

    async def refresh(self) -> NoReturn:
        """Retrieve the latest values from the API."""
        raise NotImplementedError

    async def save(self) -> NoReturn:
        """Update or create the entity."""
        raise NotImplementedError
