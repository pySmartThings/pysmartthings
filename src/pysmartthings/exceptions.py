"""Asynchronous Python client for SmartThings."""


class SmartThingsError(Exception):
    """Generic exception."""


class SmartThingsConnectionError(SmartThingsError):
    """SmartThings connection exception."""


class SmartThingsAuthenticationFailedError(SmartThingsError):
    """SmartThings authentication failed exception."""


class SmartThingsNotFoundError(SmartThingsError):
    """SmartThings not found exception."""


class SmartThingsRateLimitError(SmartThingsError):
    """SmartThings rate limit exception."""
