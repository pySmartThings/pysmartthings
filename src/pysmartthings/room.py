"""Defines the rooms module."""

from __future__ import annotations

from typing import TYPE_CHECKING

from .entity import Entity

if TYPE_CHECKING:
    from .api import Api


class Room:
    """Defines a SmartThings room."""

    def __init__(self) -> None:
        """Initialize the room."""
        self._room_id = None
        self._location_id = None
        self._name = None
        self._background_image = None

    def apply_data(self, data: dict) -> None:
        """Apply the data structure to the class."""
        self._room_id = data["roomId"]
        self._location_id = data["locationId"]
        self._name = data["name"]
        self._background_image = data["backgroundImage"]

    def to_data(self) -> dict:
        """Get a data structure representing this entity."""
        return {
            "name": self._name,
            "backgroundImage": self._background_image,
        }

    @property
    def room_id(self) -> str:
        """Get the id of the room."""
        return self._room_id

    @room_id.setter
    def room_id(self, value: str) -> None:
        """Set the id of the room."""
        self._room_id = value

    @property
    def location_id(self) -> str:
        """Get the location id the room is part of."""
        return self._location_id

    @location_id.setter
    def location_id(self, value: str) -> None:
        """Set the location id of the room."""
        self._location_id = value

    @property
    def name(self) -> str:
        """Get the name of the room."""
        return self._name

    @name.setter
    def name(self, value: str) -> None:
        """Set the name of the room."""
        self._name = value

    @property
    def background_image(self) -> str:
        """Get the background image of the room."""
        return self._background_image

    @background_image.setter
    def background_image(self, value: str) -> None:
        """Set the background image."""
        self._background_image = value


class RoomEntity(Room, Entity):
    """Defines a SmartThings room entity."""

    def __init__(
        self,
        api: Api,
        data: dict | None = None,
        *,
        location_id: str | None = None,
        room_id: str | None = None,
    ) -> None:
        """Initialize the room."""
        Entity.__init__(self, api)
        Room.__init__(self)
        if data:
            self.apply_data(data)
        if location_id:
            self._location_id = location_id
        if room_id:
            self._room_id = room_id

    async def refresh(self) -> None:
        """Refresh the room."""
        data = await self._api.get_room(self._location_id, self._room_id)
        if data:
            self.apply_data(data)

    async def save(self) -> None:
        """Save changes to the room."""
        data = await self._api.update_room(
            self._location_id, self._room_id, self.to_data()
        )
        if data:
            self.apply_data(data)
