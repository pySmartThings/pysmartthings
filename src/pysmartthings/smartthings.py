"""Define the SmartThings Cloud API."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
import logging
from typing import TYPE_CHECKING, Any, Callable, Self

from aiohttp import ClientSession
from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST, METH_PUT
from yarl import URL

from .const import API_BASE
from .exceptions import SmartThingsCommandError, SmartThingsConnectionError
from .models import (
    Attribute,
    BaseLocation,
    Capability,
    Command,
    Device,
    DeviceResponse,
    DeviceStatus,
    ErrorResponse,
    Location,
    LocationResponse,
    Room,
    RoomResponse,
    Scene,
    SceneResponse,
    Status,
)

if TYPE_CHECKING:
    from collections.abc import Awaitable

_LOGGER = logging.getLogger(__name__)


@dataclass
class SmartThings:
    """Define a class for interacting with the SmartThings Cloud API."""

    request_timeout: int = 10
    _close_session: bool = False
    _token: str | None = None
    session: ClientSession | None = None
    refresh_token_function: Callable[[], Awaitable[str]] | None = None

    async def refresh_token(self) -> None:
        """Refresh token with provided function."""
        if self.refresh_token_function:
            self._token = await self.refresh_token_function()

    def authenticate(self, token: str) -> None:
        """Authenticate the user with a token."""
        self._token = token

    async def _request(
        self,
        method: str,
        uri: str,
        *,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Handle a request to SmartThings."""
        url = URL.build(
            scheme="https",
            host=API_BASE,
            port=443,
        ).joinpath(f"v1/{uri}")

        await self.refresh_token()

        headers = {
            "Accept": "application/json, text/plain, */*",
            "Authorization": f"Bearer {self._token}",
        }

        if self.session is None:
            self.session = ClientSession()
            self._close_session = True

        try:
            async with asyncio.timeout(self.request_timeout):
                response = await self.session.request(
                    method,
                    url,
                    headers=headers,
                    json=data,
                    params=params,
                )
        except asyncio.TimeoutError as exception:
            msg = "Timeout occurred while connecting to SmartThings"
            raise SmartThingsConnectionError(msg) from exception

        text = await response.text()

        if response.status == 422:
            raise SmartThingsCommandError(ErrorResponse.from_json(text))

        return text

    async def _get(self, uri: str, params: dict[str, Any] | None = None) -> str:
        """Handle a GET request to SmartThings."""
        return await self._request(METH_GET, uri, params=params)

    async def _post(
        self,
        uri: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Handle a POST request to SmartThings."""
        return await self._request(METH_POST, uri, data=data, params=params)

    async def _put(
        self,
        uri: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Handle a PUT request to SmartThings."""
        return await self._request(METH_PUT, uri, data=data, params=params)

    async def _delete(
        self,
        uri: str,
        data: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Handle a DELETE request to SmartThings."""
        return await self._request(METH_DELETE, uri, data=data, params=params)

    async def get_locations(self) -> list[BaseLocation]:
        """Retrieve SmartThings locations."""
        resp = await self._get("locations")
        return LocationResponse.from_json(resp).items

    async def get_location(self, location_id: str) -> Location:
        """Retrieve a location with the specified ID."""
        resp = await self._get(f"locations/{location_id}")
        return Location.from_json(resp)

    async def get_rooms(self, location_id: str) -> list[Room]:
        """Retrieve a list of rooms for a location."""
        resp = await self._get(f"locations/{location_id}/rooms")
        return RoomResponse.from_json(resp).items

    async def get_room(self, location_id: str, room_id: str) -> Room:
        """Retrieve a specific room."""
        resp = await self._get(f"locations/{location_id}/rooms/{room_id}")
        return Room.from_json(resp)

    async def get_devices(
        self,
        *,
        capabilities: list[Capability] | None = None,
        location_ids: list[str] | None = None,
        device_ids: list[str] | None = None,
    ) -> list[Device]:
        """Retrieve SmartThings devices."""
        params = {}
        if capabilities:
            params["capability"] = ",".join(capabilities)
        if location_ids:
            params["locationId"] = ",".join(location_ids)
        if device_ids:
            params["deviceId"] = ",".join(device_ids)
        resp = await self._get("devices", params=params)
        return DeviceResponse.from_json(resp).items

    async def get_device(self, device_id: str) -> Device:
        """Retrieve a device with the specified ID."""
        resp = await self._get(f"devices/{device_id}")
        return Device.from_json(resp)

    async def get_scenes(self, location_id: str | None = None) -> list[Scene]:
        """Retrieve SmartThings scenes."""
        params = {}
        if location_id:
            params["locationId"] = location_id
        resp = await self._get("scenes", params=params)
        return SceneResponse.from_json(resp).items

    async def execute_scene(self, scene_id: str) -> None:
        """Execute the scene with the specified ID."""
        await self._post(f"scenes/{scene_id}/execute")

    async def get_device_status(
        self, device_id: str
    ) -> dict[str, dict[Capability, dict[Attribute, Status]]]:
        """Retrieve the status of a device."""
        resp = await self._get(f"devices/{device_id}/status")
        return DeviceStatus.from_json(resp).components

    async def execute_device_command(
        self,
        device_id: str,
        capability: Capability,
        command: Command,
        component: str = "main",
        argument: int | str | list[Any] | dict[str, Any] | None = None,
    ) -> None:
        """Execute a command on a device."""
        command_payload: dict[str, Any] = {
            "component": component,
            "capability": capability,
            "command": command,
        }
        if argument is not None:
            command_payload["arguments"] = (
                argument if isinstance(argument, list) else [argument]
            )
        _LOGGER.debug("Executing command for device %s: %s", device_id, command_payload)
        response = await self._post(
            f"devices/{device_id}/commands",
            data={"commands": [command_payload]},
        )
        _LOGGER.debug("Command response: %s", response)

    async def close(self) -> None:
        """Close open client session."""
        if self.session and self._close_session:
            await self.session.close()

    async def __aenter__(self) -> Self:
        """Async enter.

        Returns
        -------
            The SmartThings object.

        """
        return self

    async def __aexit__(self, *_exc_info: object) -> None:
        """Async exit.

        Args:
        ----
            _exc_info: Exec type.

        """
        await self.close()
