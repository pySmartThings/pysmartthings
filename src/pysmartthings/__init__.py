"""A python library for interacting with the SmartThings cloud API."""

from .app import (
    APP_TYPE_LAMBDA,
    APP_TYPE_WEBHOOK,
    CLASSIFICATION_AUTOMATION,
    App,
    AppEntity,
    AppOAuth,
    AppOAuthClient,
    AppOAuthClientEntity,
    AppOAuthEntity,
    AppSettings,
    AppSettingsEntity,
)
from .capability import (
    ATTRIBUTES,
    CAPABILITIES,
    CAPABILITIES_TO_ATTRIBUTES,
    Attribute,
    Capability,
)
from .device import (
    DEVICE_TYPE_DTH,
    DEVICE_TYPE_ENDPOINT_APP,
    DEVICE_TYPE_OCF,
    DEVICE_TYPE_UNKNOWN,
    DEVICE_TYPE_VIPER,
    Command,
    Device,
    DeviceEntity,
    DeviceStatus,
    DeviceStatusBase,
)
from .errors import APIErrorDetail, APIInvalidGrantError, APIResponseError
from .installedapp import (
    InstalledApp,
    InstalledAppEntity,
    InstalledAppStatus,
    InstalledAppType,
)
from .location import Location, LocationEntity
from .oauthtoken import OAuthToken
from .room import Room, RoomEntity
from .scene import Scene, SceneEntity
from .smartthings import SmartThings
from .subscription import SourceType, Subscription, SubscriptionEntity

__all__ = [
    "APP_TYPE_LAMBDA",
    "APP_TYPE_WEBHOOK",
    "ATTRIBUTES",
    "CAPABILITIES",
    "CAPABILITIES_TO_ATTRIBUTES",
    "CLASSIFICATION_AUTOMATION",
    "DEVICE_TYPE_DTH",
    "DEVICE_TYPE_ENDPOINT_APP",
    "DEVICE_TYPE_OCF",
    "DEVICE_TYPE_UNKNOWN",
    "DEVICE_TYPE_VIPER",
    "APIErrorDetail",
    "APIInvalidGrantError",
    "APIResponseError",
    "App",
    "AppEntity",
    "AppOAuth",
    "AppOAuthClient",
    "AppOAuthClientEntity",
    "AppOAuthEntity",
    "AppSettings",
    "AppSettingsEntity",
    "Attribute",
    "Capability",
    "Command",
    "Device",
    "DeviceEntity",
    "DeviceStatus",
    "DeviceStatusBase",
    "InstalledApp",
    "InstalledAppEntity",
    "InstalledAppStatus",
    "InstalledAppType",
    "Location",
    "LocationEntity",
    "OAuthToken",
    "Room",
    "RoomEntity",
    "Scene",
    "SceneEntity",
    "SmartThings",
    "SourceType",
    "Subscription",
    "SubscriptionEntity",
]
