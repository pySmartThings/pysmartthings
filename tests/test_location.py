"""Tests for the Location module."""

from aiohttp.hdrs import METH_GET
from aioresponses import aioresponses
from syrupy import SnapshotAssertion

from pysmartthings import SmartThings
from . import load_fixture

from .const import MOCK_URL, HEADERS


async def test_fetching_all_locations(
    client: SmartThings,
    responses: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting all locations."""
    responses.get(
        f"{MOCK_URL}/locations", status=200, body=load_fixture("locations.json")
    )
    assert await client.get_locations() == snapshot
    responses.assert_called_once_with(
        f"{MOCK_URL}/locations",
        METH_GET,
        headers=HEADERS,
        params=None,
        json=None,
    )


async def test_fetching_single_location(
    client: SmartThings,
    responses: aioresponses,
    snapshot: SnapshotAssertion,
) -> None:
    """Test getting a single location."""
    responses.get(
        f"{MOCK_URL}/locations/397678e5-9995-4a39-9d9f-ae6ba310236b",
        status=200,
        body=load_fixture("location.json"),
    )
    assert await client.get_location("397678e5-9995-4a39-9d9f-ae6ba310236b") == snapshot
    responses.assert_called_once_with(
        f"{MOCK_URL}/locations/397678e5-9995-4a39-9d9f-ae6ba310236b",
        METH_GET,
        headers=HEADERS,
        params=None,
        json=None,
    )
