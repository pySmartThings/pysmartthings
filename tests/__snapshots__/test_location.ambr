# serializer version: 1
# name: test_fetching_all_locations
  list([
    dict({
      'location_id': '397678e5-9995-4a39-9d9f-ae6ba310236b',
      'name': 'Home',
    }),
    dict({
      'location_id': '5c03e518-118a-44cb-85ad-7877d0b302e4',
      'name': 'Test Home',
    }),
  ])
# ---
# name: test_fetching_single_location
  dict({
    'country_code': 'USA',
    'latitude': 45.00708112,
    'location_id': '397678e5-9995-4a39-9d9f-ae6ba310236b',
    'longitude': -93.11223629,
    'name': 'Test Home',
    'region_radius': 150,
    'temperature_scale': 'F',
  })
# ---
